keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,233 @@
+---
+kep-number: 0
+title: kubernetes-csi release process
+authors:
+  - "@pohly"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-testing
+  - sig-release
+reviewers:
+  - @msau42
+approvers:
+  - @msau42
+  - sig-testing: TBD
+  - sig-release: TBD
+editor: @pohly
+creation-date: 2019-02-04
+last-updated: 2019-02-04
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# kubernetes-csi-release-process
+
+## Table of Contents
+
+- [kubernetes-csi-release-process](#kubernetes-csi-release-process)
+    - [Table of Contents](#table-of-contents)
+    - [Summary](#summary)
+    - [Motivation](#motivation)
+        - [Goals](#goals)
+        - [Non-Goals](#non-goals)
+    - [Proposal](#proposal)
+        - [Versionioning](#versionioning)
+            - [Release artifacts](#release-artifacts)
+            - [Release process](#release-process)
+        - [Implementation Details](#implementation-details)
+        - [Risks and Mitigations](#risks-and-mitigations)
+    - [Graduation Criteria](#graduation-criteria)
+    - [Implementation History](#implementation-history)
+    - [Drawbacks](#drawbacks)
+    - [Alternatives](#alternatives)
+    - [Infrastructure Needed](#infrastructure-needed)
+
+
+## Summary
+
+The Kubernetes [Storage
+SIG](https://github.com/kubernetes/community/tree/master/sig-storage)
+maintains a set of components under the
+[kubernetes-csi](https://github.com/kubernetes-csi) GitHub
+organization. Those components are intentionally not part of core
+Kubernetes even though they are maintained by the Kubernetes project.
+
+This document explains how these components are released.
+
+## Motivation
+
+So far, the process for tagging and building components has been
+fairly manual, with some assistance by Travis CI. There has been no
+automatic end-to-end testing.
+
+With CSI support reaching GA in Kubernetes 1.13, it is time to define
+and follow a less work-intensive and more predictable approach.
+
+### Goals
+
+- define what a "kubernetes-csi release" is
+- unit testing as mandatory pre-submit check for all components 
+- E2E testing as mandatory pre-submit check for components that get
```

Comment:
Can you describe more details of the e2e testing further down?  Specifically addressing these questions:

* Are these Kubernetes e2es with a mock/hostpath driver?  Or is it just that sidecar + a driver?
* If it's a full e2e with other sidecars, what versions do we use for those sidecars?

Reply:
I think it's fine to just support one combination to keep our testing matrix simple.  However, I think it would be good to have the ability to keep the driver and other sidecar versions stable, while only keeping the current sidecar being tested as a canary/pull.  That way failures in other sidecars won't block.

That may not work though if some change requires multiple sidecars and/or driver to be updated at the same time

Reply:
I really need to spell this out in the docs, but yes, that's exactly what I meant with "current content of kubernetes-csi release".

The "some change requires multiple sidecars and/or driver to be updated at the same time" case is already listed as a drawback of this approach.

Reply:
Maybe we need to look into adding integration tests instead of full e2e tests so that we can test sidecars individually, ie just the K8s API server, 1 sidecar, and a mock driver.

We could still have a full e2e CI Job with the hostpath driver testing canaries on all the sidecars

Reply:
How about this:
- each component runs inside the same Prow environment where it can push images and bring up and use a test cluster
- each component has its own definition what kind of tests it runs inside that test cluster
- we share the same kind of E2E testing in all components via csi-release-tools (install hostpath driver, run storage testsuite), but each component needs to opt into that and can configure that common testing (select relevant tests, influence deployment); when running these tests, the other sidecars are the ones defined by the csi-driver-host-path example, only the sidecar from the current component is the one that is being tested
- this shared E2E testing can be augmented or replaced with new per-component integration tests

This way, if there is change in a sidecar that affects the example deployment, the deployment can be temporarily defined inside the component until we are through with the transition.

Reply:
This sounds good to me.